src/random_lcg.C

```diff
@@ -0,0 +1,144 @@
+  return prn_seed[stream] * prn_norm;
+}
+
+//==============================================================================
+// FUTURE_PRN
+//==============================================================================
+
+extern "C" double
+future_prn(int64_t n)
+{
+  return future_seed(static_cast<uint64_t>(n), prn_seed[stream]) * prn_norm;
+}
+
+//==============================================================================
+// SET_PARTICLE_SEED
+//==============================================================================
+
+extern "C" void
+set_particle_seed(int64_t id)
+{
+  for (int i = 0; i < N_STREAMS; i++) {
+    prn_seed[i] = future_seed(static_cast<uint64_t>(id) * prn_stride, seed + i);
+  }
+}
+
+//==============================================================================
+// ADVANCE_PRN_SEED
+//==============================================================================
+
+extern "C" void
+advance_prn_seed(int64_t n)
+{
+  prn_seed[stream] = future_seed(static_cast<uint64_t>(n), prn_seed[stream]);
+}
+
+//==============================================================================
+// FUTURE_SEED
+//==============================================================================
+
+uint64_t
+future_seed(uint64_t n, uint64_t seed)
+{
+  // Make sure nskip is less than 2^M.
+  n &= prn_mask;
+
+  // The algorithm here to determine the parameters used to skip ahead is
+  // described in F. Brown, "Random Number Generation with Arbitrary Stride,"
+  // Trans. Am. Nucl. Soc. (Nov. 1994). This algorithm is able to skip ahead in
+  // O(log2(N)) operations instead of O(N). Basically, it computes parameters G
+  // and C which can then be used to find x_N = G*x_0 + C mod 2^M.
+
+  // Initialize constants
+  uint64_t g     {prn_mult};
+  uint64_t c     {prn_add};
+  uint64_t g_new {1};
+  uint64_t c_new {0};
+
+  while (n > 0) {
+    // Check if the least significant bit is 1.
+    if (n & 1) {
+      g_new *= g;
+      c_new = c_new * g + c;
+    }
+    c *= (g + 1);
+    g *= g;
+
+    // Move bits right, dropping least significant bit.
+    n >>= 1;
+  }
+
+  // With G and C, we can now find the new seed.
+  return (g_new * seed + c_new) & prn_mask;
+}
+
+//==============================================================================
+// PRN_SET_STREAM
+//==============================================================================
+
+extern "C" void
+prn_set_stream(int i)
+{
+  stream = i;  // Shift by one to move from Fortran to C indexing.
+}
+
+//==============================================================================
+//                               API FUNCTIONS
+//==============================================================================
+
+extern "C" int
+openmc_set_seed(int64_t new_seed)
+{
+  seed = new_seed;
+#pragma omp parallel
+  for (int i = 0; i < N_STREAMS; i++) {
+    prn_seed[i] = seed + i;
+  }
+  prn_set_stream(STREAM_TRACKING);
+#pragma end omp parallel
```

Comment:
There actually is no `#pragma end omp parallel` for C/C++. The correct way to do this is to put the relevant code in a compound statement:
```C++
#pragma omp parallel
  {
    for (int i = 0; i < N_STREAMS; i++) {
      prn_seed[i] = seed + i;
    }
    prn_set_stream(STREAM_TRACKING);
  }
```